[Pytorch] Enable autovec on aarch64 for type conversion

Nicoshev opened the pull request:
Summary:
Implementing autovec template for type conversions on aarch64-NEON

Generated code can be seen here: https://godbolt.org/z/1K6T1d9TE

We've seen significant performance improvements for converting to and from bytes, compiling using clang with -march=armv9-a+sve2:

Before
float->uint8->float ===> 683.212us
float->int8->float ===> 687.846us
int32->uint8->int32 ===> 497.121us
int32->int8->int32 ===> 481.889us

After:
float->uint8->float ===> 198.204us  ----> 245% higher throughput
float->int8->float ===> 200.241us ----> 244% higher throughput
int32->uint8->int32 ===> 197.970us ----> 151% higher throughput
int32->int8->int32 ===> 198.206us ----> 143% higher throughput

Test Plan:

buck2 test mode/opt //caffe2/test:test_ops
buck2 test mode/opt //caffe2/test:torch

Differential Revision: D85213420




cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @snadampal @milpuz01 @nikhil-arm @fadara01

Nicoshev commented:
@pytorchbot label "topic: not user facing"

MatzeB commented on aten/src/ATen/cpu/vec/vec128/vec128_convert.h:
```diff
@@ -5,6 +5,117 @@
+    return convertImpl<from_type, to_type>(src, dst, n);               \
+  }
+
+CONVERT_TEMPLATE(uint8_t, uint8_t)
```
By my reading`convertImpl` mainly does two things:
- Adding `restrict` qualifiers to the arguments. Let's hope this is always true. Though the old manually vectorized code made the same assumption, so I guess that's not really new here...

- The `memcpy` seem superfluous to me here. Seems to be a pattern to avoid strict aliasing problems, but I don't see any indication here that we have to worry about that...

MatzeB replied:
Generally this code is simpler/easier and leaves the compiler more freedom to vectorize than the previous manually vectorized code. So it's mostly a task of proving with benchmarks that the vectorization is happening? Code looks nicer in my eyes (disregarding the `memcpy` stuff)

malfet replied:
How this new `convertImpl` are different from vanialla implementation of `convert` which should do the same scalar operation, isn't it?

Nicoshev commented:
@malfet 
> @Nicoshev for your godbolt example, you are compiling it for SVE, but replacing the NEON code, so I think it would be fair to see the output generated for vectorized for those and compared to the code generated before

New code: https://godbolt.org/z/oY6s9K5sM
Old code: https://godbolt.org/z/h9n6czqdz

Code size is greatly reduced. This won't have an impact on benchmarks, but it will affect prod positively

malfet commented on aten/src/ATen/cpu/vec/vec128/vec128_convert.h:
```diff
@@ -5,6 +5,114 @@
+  for (uint64_t i = 0; i < len; i++) {
+    dst[i] = static_cast<to_type>(src[i]);
+  }
+}
```
This is exactly identical to the default implementation of convert template, isn't it?

Nicoshev replied:
@malfet no, it adds the restrict qualifiers and casts the count as uint64

malfet commented on aten/src/ATen/cpu/vec/vec128/vec128_convert.h:
```diff
@@ -5,6 +5,114 @@
+CONVERT_TEMPLATE(int16_t, int32_t)
+CONVERT_TEMPLATE(int16_t, int64_t)
+CONVERT_TEMPLATE(int16_t, float)
+CONVERT_TEMPLATE(int16_t, double)
```
It's very easy to make mistake in this endless repetition, any way it could be refactored further?  (for example, it looks like you've missed float16<->bfloat16 conversions in this list don't you?)

Nicoshev committed:
[Pytorch] Enable autovec on aarch64 for type conversion (#166049)

Summary:

Implementing autovec template for type conversions on aarch64-NEON

We've seen significant performance improvements for converting to and from bytes:

Before
float->uint8->float ===> 683.212us
float->int8->float ===> 687.846us
int32->uint8->int32 ===> 497.121us
int32->int8->int32 ===> 481.889us

After:
float->uint8->float ===> 198.204us  ----> 245% higher throughput
float->int8->float ===> 200.241us ----> 244% higher throughput
int32->uint8->int32 ===> 197.970us ----> 151% higher throughput
int32->int8->int32 ===> 198.206us ----> 143% higher throughput

Test Plan:
Correctness:

buck2 test mode/opt //caffe2/test:test_ops
buck2 test mode/opt //caffe2/test:torch

Performance:

buck2 run mode/opt //caffe2/benchmarks/operator_benchmark/fb:operator_benchmark_test

Reviewed By: mcfi

Differential Revision: D85213420

malfet reviewed (changes requested):
PR title claims throughput improvements, but does not update any benchmark files. 

Moreover, `convertImpl` feels just like a code duplicate of default template implementation already, isn't it?

Nicoshev commented:
@pytorchbot merge -f "Benchmark failures are a pre-existing issue"

Nicoshev commented:
@ezyang Issue got fixed in https://github.com/pytorch/pytorch/pull/166739
